### ISSUE TITLE
type(self)() in inherited method inferred as parent class (E1101)

### ISSUE THREAD

**Issue by zuntrax** [PyCQA/pylint#2329]
### Steps to reproduce
1. Create `test.py`:
```python
#!/usr/bin/env python3

# pylint: disable=too-few-public-methods,missing-docstring

class Path:
    def __init__(self, name):
        self.name = name

    def __getitem__(self, name):
        return type(self)(f"{self.name}/{name}")

class Special(Path):
    def action(self):
        print(self.name)


def main():
    spam = Special("spam")
    eggs = spam["eggs"]
    print(type(eggs))
    eggs.action()


if __name__ == "__main__":
    main()
```
2. Run pylint.
```
pylint test.py
```

### Current behavior
```
************* Module test
test.py:21:4: E1101: Instance of 'Path' has no 'action' member (no-member)

------------------------------------------------------------------
Your code has been rated at 6.67/10 (previous run: 6.67/10, +0.00)
```

### Expected behavior
No finding. `eggs` should have been inferred as instance of `Special`, not `Path`.

### pylint --version output
```
pylint 2.0.0
astroid 2.0
Python 3.6.6 (default, Jun 27 2018, 13:11:40) 
[GCC 8.1.1 20180531]
```

**Comment by brycepg** [comment]
Thanks for the report

**Comment by brycepg** [comment]
Very strange behavior -- I can only reproduce inside the `visit_attribute` method of a pylint Checker. Direct astroid inference via extract_node works fine.

**Comment by jacobtylerwalls** [comment]
Duplicate of https://github.com/PyCQA/pylint/issues/981 and https://github.com/PyCQA/astroid/issues/457. I'll mention this test case there.
